### PR TITLE
feat(crypto-js): Implement key verification

### PR DESCRIPTION
**Pull request by Hywan**
Related to https://github.com/matrix-org/matrix-rust-sdk/issues/1016.

This PR is quite important. It brings the Key Verification API to `crypto-js`, aka SAS, Emoji, QR Code and bootstraping cross stuff.

## Progression

* [x] Support `OlmMachine.get_user_devices` (imply supporting a `UserDevices` type),
* [x] Support `OlmMachine.get_device` (imply supporting a `Device` type),
* [x] Support `OlmMachine.get_verification` (imply supporting a `Verificaftion` type),
* [x] Support `OlmMachine.get_verification_request` and `.get_verification_requests` (imply supporting a `VerificationRequest` type),
* [x] Support `OlmMachine.receive_unencrypted_verification_event`,
* [x] Implement `Verification`,
* [x] Implement `VerificationRequest`,
* [x] Implement `Sas`,
* [x] Implement `Emoji`
* [x] Implement `Qr`,
* [x] Implement `QrCode`,
* [x] Implement `CancelInfo`,
* [x] Implement `CancelCode`,
* [x] Implement `DeviceKey`,
* [x] Implement `Device`,
* [x] Implement `LocalTrust`,
* [x] Implement `UserDevices`.

### CODE REVIEW

File: bindings/matrix-sdk-crypto-js/src/device.rs [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+    /// The device has been verified and is trusted.
+    Verified,
+
+    /// The device been blacklisted from communicating.
+    BlackListed,
+
+    /// The trust state of the device is being ignored.
+    Ignored,
+
+    /// The trust stte is unset.
```
Comment (gnunicorn) [inline bindings/matrix-sdk-crypto-js/src/device.rs:194]:
```suggestion
    /// The trust state is unset.
```

File: bindings/matrix-sdk-crypto-js/src/requests.rs [hunk 2]
```diff
@@ -256,12 +261,33 @@ pub struct KeysBackupRequest {
...
+#[derive(Debug)]
+pub struct SigningKeysUploadRequest {
+    /// The request ID.
+    #[wasm_bindgen(readonly)]
+    pub id: Option<JsString>,
+
+    /// A JSON-encoded object of form:
+    ///
+    /// ```json
+    /// {"master_key", …, "self_signing_key": …, "user_signing_key": …}
```
Comment (gnunicorn) [inline bindings/matrix-sdk-crypto-js/src/requests.rs:279]: that's not correct json ... but with `...` that never will be ...
```suggestion
    /// {"master_key": …, "self_signing_key": …, "user_signing_key": …}
```

File: bindings/matrix-sdk-crypto-js/tests/helper.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
+const { DeviceLists, RequestType, KeysUploadRequest, KeysQueryRequest } = require('../pkg/matrix_sdk_crypto_js');
+
+function* zip(...arrays) {
```
Comment (gnunicorn) [inline bindings/matrix-sdk-crypto-js/tests/helper.js:3]: what's that `function*` notation? What does that mean? Not seen that in JS before!?!

**Review by gnunicorn — APPROVED** [review]
Not super familiar with the code so I mostly looked for consistency, documentation and typos. Found minor ones, but I think this is fine overall.

File: bindings/matrix-sdk-crypto-js/src/device.rs (hunk 1 above)
Comment (Hywan) [reply bindings/matrix-sdk-crypto-js/src/device.rs:194]: Thanks!

File: bindings/matrix-sdk-crypto-js/src/requests.rs (hunk 2 above)
Comment (Hywan) [reply bindings/matrix-sdk-crypto-js/src/requests.rs:279]: ;-)

File: bindings/matrix-sdk-crypto-js/tests/helper.js (hunk 3 above)
Comment (Hywan) [reply bindings/matrix-sdk-crypto-js/tests/helper.js:3]: It's used to declare a [`Generator`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator), with `yield` to produce a value.

**Comment by Hywan** [comment]
And a special thanks to @poljar for the help with the unit tests!

**Pull request merged**